README.md
What's ?
===============
chef で使用する PHP の cookbook です。

Usage
-----
cookbook なので berkshelf で取ってきて使いましょう。

* Berksfile
```ruby
source "https://supermarket.chef.io"

cookbook "php", git: "https://github.com/bageljp/cookbook-php.git"
```

```
berks vendor
```

#### Role and Environment attributes

* sample_role.rb
```ruby
override_attributes(
  "php" => {
    "install_flavor" => "yum"
  }
)
```

Recipes
----------

#### php::default
phpのインストールと設定。

#### php::fpm
php-fpmの設定。

#### php::opcache
opcacheのインストールと設定。

Attributes
----------

主要なやつのみ。

#### php::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><tt>['php']['rpm']['url']</tt></td>
    <td>string</td>
    <td>rpmでインストールする場合にrpmが置いてあるURL。rpmbuildしたものをs3とかに置いておくといいかも。</td>
  </tr>
  <tr>
    <td><tt>['php']['user']</tt></td>
    <td>string</td>
    <td>php-fpmの起動ユーザ。</td>
  </tr>
</table>



CHANGELOG.md
php CHANGELOG
=============

This file is used to list changes made in each version of the php cookbook.

0.1.0
-----
- [kadoyama] - Initial release of php

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/fpm.rb
#
# Cookbook Name:: php
# Recipe:: fpm
#
# Copyright 2014, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#

case node['platform']
when 'amazon'
  case node['php']['version']['major']
  when '5.5'
    %w(
      php55-fpm
    ).each do |pkg|
      package pkg do
        action :install
      end
    end

    %w(
      php-5.5.ini
      php-fpm-5.5.conf
      php-fpm-5.5.d/php-fpm.conf
      php-fpm-5.5.d/www.conf
    ).each do |t|
      template "/etc/#{t}" do
        owner "root"
        group "root"
        mode 00644
        source "#{t}.erb"
        notifies :restart, "service[php-fpm-#{node['php']['version']['major']}]"
      end
    end

    template "/etc/logrotate.d/php-fpm-#{node['php']['version']['major']}" do
      owner "root"
      group "root"
      mode 00644
      source "logrotate.php-fpm-#{node['php']['version']['major']}.erb"
    end

    directory "/var/log/php-fpm/#{node['php']['version']['major']}" do
      owner "#{node['php']['fpm']['user']}"
      group "root"
      mode 00770
    end

    file "/var/log/php-fpm/#{node['php']['version']['major']}/www-error.log" do
      owner "#{node['php']['fpm']['user']}"
      group "#{node['php']['fpm']['group']}"
       mode 00644
    end

    service "php-fpm-#{node['php']['version']['major']}" do
      supports :status => true, :restart => true, :reload => true
      action [ :enable, :start ]
    end
  end
end


recipes/opcache.rb
#
# Cookbook Name:: php
# Recipe:: opcache
#
# Copyright 2014, bageljp
#
# All rights reserved - Do Not Redistribute
#

case node['platform']
when "amazon"
  package "php55-opcache"
when "redhat","centos"
  case node['php']['version']['major']
  when '5.5'
    case node['php']['install_flavor']
    when 'yum'
      %w(
        php55u-opcache
      ).each do |pkg|
        package pkg do
          action :install
          options "--enablerepo=ius"
        end
      end
    when 'rpm'
      rpms = [
        "php55u-opcache"
      ]
      rpms.each do |pkg|
        remote_file "/usr/local/src/#{pkg}-#{node['php']['version']['major']}.#{node['php']['version']['minor']}-1.ius.el6.x86_64.rpm" do
          owner "root"
          group "root"
          mode 00644
          source "#{node['php']['rpm']['url']}#{pkg}-#{node['php']['version']['major']}.#{node['php']['version']['minor']}-1.ius.el6.x86_64.rpm"
        end
      end

      rpms_join = rpms.join(" ")
      rpms_join_pkg = ""
      rpms.each do |pkg|
        rpms_join_pkg += "#{pkg}-#{node['php']['version']['major']}.#{node['php']['version']['minor']}-1.ius.el6.x86_64.rpm "
      end
      bash "install php" do
        user "root"
        cwd "/usr/local/src"
        code <<-EOC
          rpm -Uvh #{rpms_join_pkg}
        EOC
        not_if "rpm -q #{rpms_join} > /dev/null 2>&1"
      end
    end
  end
end


recipes/default.rb
#
# Cookbook Name:: php
# Recipe:: default
#
# Copyright 2014, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#

case node['platform']
when "amazon"
  case node['php']['version']['major']
  when '5.5'
    %w(
      php55
      php55-devel
      php55-mbstring
      php55-mysqlnd
    ).each do |pkg|
      package pkg do
        action :install
      end
    end

    template "/etc/php-#{node['php']['version']['major']}.ini" do
      owner "root"
      group "root"
      mode 00644
      source "php-#{node['php']['version']['major']}.ini.erb"
      unless node['php']['webserver'].nil?
        notifies :restart, "service[#{node['php']['webserver']}]"
      end
    end
  end
when "redhat","centos"
  include_recipe "yum-ius::default"

  case node['php']['version']['major']
  when '5.5'
    case node['php']['install_flavor']
    when 'yum'
      %w(
        php55u
        php55u-devel
        php55u-mbstring
        php55u-mysqlnd
      ).each do |pkg|
        package pkg do
          action :install
          options "--enablerepo=ius"
        end
      end
    when 'rpm'
      %w(
        pcre-devel
        libXpm-devel
        libxslt
        libxslt-devel
      ).each do |pkg|
        package pkg
      end

      case node['platform']
      when "redhat"
        %w(
          t1lib
          libvpx
        ).each do |pkg|
          package pkg
        end
      when "centos"
        %w(
          t1lib-devel
          libvpx-devel
        ).each do |pkg|
          package pkg
        end
      end

      rpms = [
        "php55u",
        "php55u-cli",
        "php55u-common",
        "php55u-devel",
        "php55u-gd",
        "php55u-mbstring",
        "php55u-pdo",
        "php55u-mysqlnd",
        "php55u-process",
        "php55u-xml"
      ]
      rpms.each do |pkg|
        remote_file "/usr/local/src/#{pkg}-#{node['php']['version']['major']}.#{node['php']['version']['minor']}-1.ius.el6.x86_64.rpm" do
          owner "root"
          group "root"
          mode 00644
          source "#{node['php']['rpm']['url']}#{pkg}-#{node['php']['version']['major']}.#{node['php']['version']['minor']}-1.ius.el6.x86_64.rpm"
        end
      end

      rpms_join = rpms.join(" ")
      rpms_join_pkg = ""
      rpms.each do |pkg|
        rpms_join_pkg += "#{pkg}-#{node['php']['version']['major']}.#{node['php']['version']['minor']}-1.ius.el6.x86_64.rpm "
      end
      bash "install php" do
        user "root"
        cwd "/usr/local/src"
        code <<-EOC
          yumdownloader php55u-pear php55u-pecl-jsonc php55u-pecl-jsonc-devel --enablerepo=ius
          rpm -Uvh #{rpms_join_pkg} php55u-pear-*.rpm php55u-pecl-jsonc-*.rpm php55u-pecl-jsonc-devel-*.rpm
        EOC
        not_if "rpm -q #{rpms_join} php55u-pear php55u-pecl-jsonc php55u-pecl-jsonc-devel > /dev/null 2>&1"
      end
    end

    template "/etc/php.ini" do
      owner "root"
      group "root"
      mode 00644
      source "php-#{node['php']['version']['major']}.ini.erb"
      unless node['php']['webserver'].nil?
        notifies :restart, "service[#{node['php']['webserver']}]"
      end
    end
  end
end
